### PR TITLE
1443164: no_proxy matches the host name when *.redhat.com is used

### PR DESCRIPTION
**Pull request by jirihnidek**
* Bug fix: https://bugzilla.redhat.com/show_bug.cgi?id=1443164
* Method proxy_bypass_environment() from urllib does not
  support no_proxy with items containing asterisk
  (e.g.: *.redhat.com). Thus values in this environment variable
  is preprocessed and leading asterisks are stripped.
* There will be opened PR for urllib to address this problem
  in proper way.

### CODE REVIEW

**Comment by jirihnidek** [comment]
[Bug 1443164 - no_proxy does not match the host name when *.redhat.com is used](https://trello.com/c/Xs1zsRxa/401-bug-1443164-no-proxy-does-not-match-the-host-name-when-redhat-com-is-used)

File: python-rhsm/src/rhsm/connection.py [hunk 1]
```diff
@@ -230,6 +230,23 @@ def _encode_auth(username, password):
...
+    This fixes no_proxy/NO_PROXY environment to not include leading
+    asterisk, because there is some imperfection in proxy_bypass_environment.
+    """
+
+    # This fixes BZ: 1443164, because proxy_bypass_environment() from urllib does
+    # not support no_proxy with items containing asterisk (e.g.: *.redhat.com)
+
+    no_proxy = os.environ.get('no_proxy') or os.environ.get('NO_PROXY')
+    # Remove all leading white spaces and asterisks from items of no_proxy
+    no_proxy = ','.join([item.lstrip(' *') for item in no_proxy.split(',')])
```
Comment (kahowell) [inline python-rhsm/src/rhsm/connection.py]: Need to handle the case of `no_proxy=None`...

**Comment by jirihnidek** [comment]
I pushed these changes to prove tests will not fail at Jenkins. I plan to add more unit tests covering new code. Please don't merge it until more unit tests are added to this PR.

**Comment by jirihnidek** [comment]
Bug report for Python created: http://bugs.python.org/issue30462

**Comment by jirihnidek** [comment]
Created right solution: PR for Python: https://github.com/python/cpython/pull/1795, but this is long term solution with uncertain outcome.

**Comment by jirihnidek** [comment]
The failure of unit tests is strange. It works at my laptop:
```bash
cd python-rhsm
nosetests test/unit
```
With following result:
```console
...
Ran 324 tests in 0.258s

OK
```
Isn't `python-rhsm` loaded from system at Jenkins build machine?

**Comment by kahowell** [comment]
@jirihnidek I can reproduce the Jenkins failures on my machine. I noticed some of our tests are a little brittle. If you've modified your /etc/rhsm/rhsm.conf, it can interfere with some of the tests. Another possibility is code that didn't get checked in.

**Comment by jirihnidek** [comment]
I tested it at two another VMs without any failure. I will investigate it more tomorrow.

**Pull request merged**